core/src/replay_stage.rs

```diff
@@ -2698,13 +2698,13 @@ impl ReplayStage {
         voted_signatures: &mut Vec<Signature>,
         epoch_slots_frozen_slots: &mut EpochSlotsFrozenSlots,
     ) {
-        bank_forks.write().unwrap().set_root(
+        let mut w_bank_forks = bank_forks.write().unwrap();
+        w_bank_forks.set_root(
             new_root,
             accounts_background_request_sender,
             highest_confirmed_root,
         );
-        let r_bank_forks = bank_forks.read().unwrap();
-        let new_root_bank = &r_bank_forks[new_root];
+        let new_root_bank = &w_bank_forks[new_root];
```

Comment:
hmm what was the race we were trying to prevent here? Replay is the only thread to ever insert/remove from `BankForks` so the contents shouldn't change. 

Ideally we would like to hold the write lock as little as possible so things like RPC don't contend with any of the remaining logic

Reply:
`bank_forks` might be modified between the time write-lock is released and the read-lock is obtained here:
https://github.com/solana-labs/solana/blob/9d3afba04/core/src/replay_stage.rs#L2701-L2706

Effectively we need an atomic downgrade here to release the write-lock and obtain the read-lock without letting any writes go in between, but that is not supported yet: https://github.com/rust-lang/rust/issues/32527

Reply:
It looks like there is only one caller running updates on this structure and is doing so from a single thread. After the update we don't really need to hold any lock. I guess we're just using the read lock to get access to the structure. Is this correct @carllin. If this is the case a comment may be in order. I just noticed this from code inspection. Without checking callers the drop/reacquire looks suspect.

Reply:
> I guess we're just using the read lock to get access to the structure
Yeah exactly only this thread ever updates BankForks, which is why it should be safe.

Reply:
I think the risk here is that one can easily introduce changes to replay, bankforks, or elsewhere which invalidates that assumption and introduce the race-condition here, without any compile-time or test errors.
We had exactly this scenario with push-votes:
https://github.com/solana-labs/solana/pull/18552#pullrequestreview-703284639

Reply:
My vote is to target for correctness and thread-safety and optimize later if it turned out to be a bottleneck. But I leave this to Carl since he knows this code much better than me.

Reply:
I'm for keeping it as is right now. 

I see Behzad's point that if we introduced the ability for another thread to write to BankForks this could break, but that would break a bunch of other places in Replay as well. 

If we ever decided we wanted to introduce a model in which more than one thread could update BankForks, I think we would have to overhaul a significant portion of ReplayStage, at which point I think i would be appropriate to go over these things.

Until then we work under the assumption that BankForks can only have single threaded updates seems fine.

Reply:
> If this is the case a comment may be in order. I just noticed this from code inspection.

@jbiseda marking a comment here is not a bad idea.

Reply:
leaving as-is